contrib/db_pools/lib/src/database.rs

```diff
@@ -0,0 +1,117 @@
+    type Target = <D::Pool as Pool>::Connection;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<D: Database> std::ops::DerefMut for Connection<D> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[rocket::async_trait]
+impl<'r, D: Database> FromRequest<'r> for Connection<D> {
+    type Error = Error<<D::Pool as Pool>::GetError>;
+
+    async fn from_request(req: &'r Request<'_>) -> Outcome<Self, Self::Error> {
+        let db: &D = match req.rocket().state() {
+            Some(p) => p,
+            _ => {
+                let dbtype = Paint::default(std::any::type_name::<D>()).bold();
+                let fairing = Paint::default(format!("{}::fairing()", dbtype)).wrap().bold();
+                error!("requesting `{}` DB connection without attaching `{}`.", dbtype, fairing);
+                info_!("Attach `{}` to use database connection pooling.", fairing);
+                return Outcome::Failure((Status::InternalServerError, Error::UnattachedFairing));
+            }
+        };
+
+        match db.pool().get().await {
+            Ok(conn) => Outcome::Success(Connection(conn)),
+            Err(e) => Outcome::Failure((Status::ServiceUnavailable, Error::Db(e))),
+        }
+    }
+}
+
+impl<D: Database> Sentinel for Connection<D> {
+    fn abort(rocket: &Rocket<Ignite>) -> bool {
+        if rocket.state::<D>().is_none() {
+            let dbtype = Paint::default(std::any::type_name::<D>()).bold();
+            let fairing = Paint::default(format!("{}::fairing()", dbtype)).wrap().bold();
+            error!("requesting `{}` DB connection without attaching `{}`.", dbtype, fairing);
+            info_!("Attach `{}` to use database connection pooling.", fairing);
+            return true;
+        }
+
+        false
+    }
+}
+
+/// The database fairing for pool types created with the `pool!` macro.
+pub struct Fairing<D: Database>(&'static str, fn(D::Pool) -> D);
+
+impl<D: Database> Fairing<D> {
+    /// Create a new database fairing with the given constructor.  This
+    /// constructor will be called to create an instance of `D` after the pool
+    /// is initialized and before it is placed into managed state.
+    pub fn new(fairing_name: &'static str, ctor: fn(D::Pool) -> D, ) -> Self {
+        Self(fairing_name, ctor)
+    }
+}
+
+#[rocket::async_trait]
+impl<D: Database> rocket::fairing::Fairing for Fairing<D> {
+    fn info(&self) -> Info {
+        Info {
+            name: self.0,
+            kind: Kind::Ignite,
+        }
+    }
+
+    async fn on_ignite(&self, rocket: Rocket<Build>) -> Result<Rocket<Build>, Rocket<Build>> {
+        let pool = match <D::Pool>::initialize(D::NAME, &rocket).await {
+            Ok(p) => p,
+            Err(e) => {
+                error!("error initializing database connection pool: {}", e);
+                return Err(rocket);
+            }
+        };
+
+        Ok(rocket.manage((self.1)(pool)))
```

Comment:
This will panic if one is already being managed. We should probably check and emit a nicer error.